src/runtime_src/tools/xclbinutil/Section.h

```diff
@@ -22,20 +22,44 @@
 // #includes here - please keep these to a bare minimum!
 #include "xclbin.h"
 
-#include <string>
+#include <boost/property_tree/ptree.hpp>
 #include <fstream>
-#include <map>
 #include <functional>
+#include <map>
+#include <memory>
+#include <string>
 #include <vector>
-
-#include <boost/property_tree/ptree.hpp>
-
-// ------------ F O R W A R D - D E C L A R A T I O N S ----------------------
-// Forward declarations - use these instead whenever possible...
-
 // ------------------- C L A S S :   S e c t i o n ---------------------------
 
 class Section {
+ typedef std::function<Section*()> Section_factory;
+
+ protected:
+  class SectionInfo {
+    SectionInfo() = delete;
+
+   public:
+    SectionInfo(enum axlf_section_kind eKind, const std::string & secondName, Section_factory sectionCtor);   
+    void setNodeName(const std::string& sNodeName);
+    void setSupportsSubSections(bool bValue);
+    void setSupportsIndexing(bool bValue);
+
+    enum axlf_section_kind getType() const;
+    const std::string &getSectionName() const;
+    Section_factory getCtor() const;
+    const std::string &getNodeName() const;
+    bool getSupportsSubSections() const;
+    bool getSupportsIndexing() const;
+
+   private:
+    enum axlf_section_kind m_eKind;       // The section enumeration value
+    std::string m_sectionName;            // Name of the section 
+    Section_factory m_sectionCtor;        // Section constructor
+    std::string m_nodeName;               // JSON node name
+    bool m_bSupportsSubSections;          // Support subsections
+    bool m_bSupportsIndexing;             // Supports indexing
+  };
+
  public:
   enum FormatType{
```

Comment:
Your c++ code  enums should be class enums?

Reply:
This change is rather large (due to scoping).  

The next submission will be refactored to use class enums.